### ISSUE TITLE
Getting a bitmap from the current WebGL canvas - solved/

### ISSUE THREAD

**Issue by keikland** [AnalyticalGraphicsInc/cesium#6845]
I needed to obtain a bitmap from arbitrary locations on the Cesium map, and managed to do this by i.a. enabling preserveDrawingBuffer and stealing access to the internal webgl context.  (I saw some references to extracting a bitmap in earlier posts, but found no solutions or hints there.) I use requestRenderMode: true, so the approach appears safe.

It would be great to see bitmap extraction as an internal feature in Cesium to avoid messing with updates generally and not have to expose the internal webgl context.  It seems to me that the recent  pre/post update/render functions could be a way to integrate this harmoniously, for any state of requestRenderMode and preserveDrawingBuffer.  

Any chance for seeing this new functionality in Cesium?  

I can contribute my solution for specifying and getting out the bitmap in 2d-compatible form (a surprisingly convoluted process, but ultimately not that much code.)  I suspect some of you Cesium rendering experts would quickly think of a n implementation...  

Kjell

PS! As an aside, I let the user select the desired bitmap rectangle/polygon by dragging the mouse.  In the process, I found that the continuously updated frame border and fill did not match the actual selection, with errors growing larger as the selection moves away from center.  I only use standard conversion functions, and the ultimate extracted bitmap is fine/as expected.  Is this a known problem?

**Comment by keikland** [comment]
I needed to obtain a bitmap from arbitrary locations on the Cesium map, and managed to do this by i.a. enabling preserveDrawingBuffer and stealing access to the internal webgl context.  (I saw some references to extracting a bitmap in earlier posts, but found no solutions or hints there.) I use requestRenderMode: true, so the approach appears safe.

It would be great to see bitmap extraction as an internal feature in Cesium to avoid messing with updates generally and not have to expose the internal webgl context.  It seems to me that the recent  pre/post update/render functions could be a way to integrate this harmoniously, for any state of requestRenderMode and preserveDrawingBuffer.  

Any chance for seeing this new functionality in Cesium?  

I can contribute my solution for specifying and getting out the bitmap in 2d-compatible form (a surprisingly convoluted process, but ultimately not that much code.)  I suspect some of you Cesium rendering experts would quickly think of a n implementation...  

Kjell

PS! As an aside, I let the user select the desired bitmap rectangle/polygon by dragging the mouse.  In the process, I found that the continuously updated frame border and fill did not match the actual selection, with errors growing larger as the selection moves away from center.  I only use standard conversion functions, and the ultimate extracted bitmap is fine/as expected.  Is this a known problem?

**Comment by hpinkos** [comment]
Hi @keikland!  This sounds like an interesting feature.  I don't think it's something that would be a priority for our team to add, but we're always happy to review a pull request if you wanted to implement this kind of functionality

**Comment by mramato** [comment]
@keikland I agree that a helper function (per perhaps just a sandcastle example) would be a good contribution to Cesium and eliminate confusion our users occasionally have.  That being said, there's no reason to mess with the webGL context or `preserveDrawingBuffer`, you just need to capture the thumbnail in the `postRender` callback:

```js
var removePostRender = scene.postRender.addEventListener(function () {
    removePostRender();
    var thumbnail = canvas.toDataURL('image/jpeg', 0.95);
});
```

As simple as the above is, there's obviously a lot of other robustness that could be added (resizing the canvas to a desired resolution, re-rendering until all tiles/data is loaded, etc... ) but there's no need for us to add every option at once.

**Comment by keikland** [comment]
Hi @mramato.  Amazing, that seems to be exactly what I was looking for, and which you seem to have tried already :-).  Obviously agree that what I did is unoptimal, but the road to discovery can be long.

The only question I have is if the canvas in the eventlistener supports getting a ctx=canvas.getContext('2d' or ctx=getContext(getContext('webgl').  

If it is '2d', I can use ctx.getImageData(), while for 'webgl' I would use ctx.readPixels() to get a specific target pixel region.  The readPixels() requires a bit of postprocessing to get it into a compatible 2d format. 

With that clarified, it should be quick to add a simple Sandcastle demo.

**Comment by bagnell** [comment]
You can't have two different contexts on the same canvas. You can read the pixels with `scene.context.readPixels`.

The other option is to create a new canvas and draw the image to it.
```javascript
var image = new HTMLImageElement();
image.src = canvas.toDataURL('image/jpeg', 0.95);
var ctx = canvas2D.getContext('2d');
ctx.drawImage(image);
```

**Comment by keikland** [comment]
Refactored my relevant logic into a single function call inside the postRender code approach by @mramato (replacing the canvas variable with scene.canvas (since I have a scene global variable, but not canvas). Maintained the preserveDrawingBuffer: true for now.

The result was Ok for scene.canvas.getContext('webgl') and readPixels(), but not '2d' with getImageData(), confirming the indirect hint by @bagnell that the underlying context is indeed 'webgl'. The toDataURL is of course supported under both context models.

I don't wish the extra steps of creating a memory/time-wasting full new canvas for a 2d context and extract a region of interest from this.  This means that my core webgl buffer-reading logic is quite Ok.   With the @mramato code, I am also happy that I do not need to access the internal webgl context variable any more. 

The next testing and improvement stage is to see if the success remains with preserveDrawingBuffer: false.  Thanks for your great help thus far.

**Comment by keikland** [comment]
Ok to close the issue @mramato, and thanks again.  The image capture solution via the webgl context is running reliably thus far, and importantly with preserveDrawingBuffer: false.  No problem with incomplete scenes since the image grabbing is user-controlled (marking a target area with the mouse), with no animation going.   

I think I can make a simple Sandcastle demo for the base logic, if it is of interest, featuring a programmatic capture. 

PS! While capturing Bing imagery would be rather boring, the implemented routine actually works with a frequently updated WMS basemap imagery source.   The capture is used both for time-line image comparison and for object comparison against a master.  It may be a fairly novel use of Cesium.
